### PR TITLE
Introduce a RoomScrollStateStore

### PR DESCRIPTION
**Pull request by dbkr**
to keep the place we're scrolled to in rooms. This mainly eleimates
the extra, superfluous onRoomViewStoreUpdate callback that
happened when the previous room saved back its scroll state.
Moving the scroll state to a separate store means we can have this
not emit events because nothing needs to know when the scroll state
changes.

### CODE REVIEW

File: src/components/structures/RoomView.js [hunk 1]
```diff
@@ -340,7 +354,7 @@ module.exports = React.createClass({
         this.unmounted = true;
 
         // update the scroll map before we get unmounted
-        this._updateScrollMap(this.state.roomId);
+        RoomScrollStateStore.setScrollState(this.state.roomId, this._getScrollState());
```
Comment (richvdh) [inline src/components/structures/RoomView.js]: `this.state.roomId` may still be `null` here, in which case we shouldn't save the scroll state.

File: src/components/structures/RoomView.js [hunk 2]
```diff
@@ -191,18 +191,32 @@ module.exports = React.createClass({
             newState.room = MatrixClientPeg.get().getRoom(newState.roomId);
         }
 
+        if (this.state.roomId !== newState.roomId) {
```
Comment (richvdh) [inline src/components/structures/RoomView.js]: this is somewhat in conflict with the comment at line 188: "This does assume that the roomID will not change for the lifetime of the RoomView instance". If we're changing room, LoggedInView gives us a new RoomView.

I think that this is hit when we resolve a room id after switching to room by alias, though, so it's sort-of valid, though I'd prefer it to be expressed more like `if (this.state.roomId === null && newState.roomId !== null)`.

File: src/components/structures/RoomView.js [hunk 3]
```diff
@@ -191,18 +191,32 @@ module.exports = React.createClass({
...
+            // Store the scroll state for the previous room so that we can return to this
+            // position when viewing this room in future.
+            if (this.state.roomId) {
+                RoomScrollStateStore.setScrollState(this.state.roomId, this._getScrollState());
```
Comment (richvdh) [inline src/components/structures/RoomView.js]: ... and this is redundant. We save the scrollState in unmount.

File: src/stores/RoomScrollStateStore.js [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+/**
+ * Stores where the user has scrolled to in each room
```
Comment (richvdh) [inline src/stores/RoomScrollStateStore.js]: Please could you describe the structure of the store? This used to be in the comments in RoomViewStore but has got lost.

File: src/stores/RoomScrollStateStore.js [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+        return this._scrollStateMap[roomId];
+    }
+
+    setScrollState(roomId, scrollState) {
+        this._scrollStateMap[roomId] = scrollState;
+    }
+}
+
+if (global.mx_RoomScrollStateStore === undefined) {
+    global.mx_RoomScrollStateStore = new RoomScrollStateStore();
```
Comment (richvdh) [inline src/stores/RoomScrollStateStore.js]: I'm not entirely convinced we want to be sticking more stuff in the global namespace?

File: src/stores/RoomScrollStateStore.js (hunk 4 above)
Comment (dbkr) [reply src/stores/RoomScrollStateStore.js]: ah, yes

File: src/components/structures/RoomView.js (hunk 2 above)
Comment (dbkr) [reply src/components/structures/RoomView.js]: yes, I think so - removed the scroll state saving here too as it'll now never be hit

File: src/components/structures/RoomView.js (hunk 3 above)
Comment (dbkr) [reply src/components/structures/RoomView.js]: oh, as per above

File: src/stores/RoomScrollStateStore.js (hunk 5 above)
Comment (dbkr) [reply src/stores/RoomScrollStateStore.js]: Unfortunately I still don't know of any other way to make a proper singleton (declaring them as module level variables gives you a different instance when accessed from a different module which causes subtle breakage when accessing between riot-web and react-sdk).

Comment (richvdh) [reply src/stores/RoomScrollStateStore.js]: https://derickbailey.com/2016/03/09/creating-a-true-singleton-in-node-js-with-es6-symbols/ suggests alternatives which are complicated. fair enough.

**Pull request merged**